transports/webrtc/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! Implementation of the [`Transport`] trait for WebRTC direct protocol. "direct" here means
+//! communicating without a signaling server.
+//!
+//! # Overview
+//!
+//! ## ICE
+//!
+//! RFCs: 8839, 8445 See also:
+//! https://tools.ietf.org/id/draft-ietf-rtcweb-sdp-08.html#rfc.section.5.2.3
+//!
+//! The WebRTC protocol uses ICE in order to establish a connection.
+//!
+//! In a typical ICE setup, there are two endpoints, called agents, that want to communicate. One
+//! of these two agents is the local browser, while the other agent is the target of the
+//! connection.
+//!
+//! Even though in this specific context all we want is a simple client-server communication, it is
+//! helpful to keep in mind that ICE was designed to solve the problem of NAT traversal.
+//!
+//! The ICE workflow works as follows:
+//!
+//! - An "offerer" determines ways in which it could be accessible (either an
+//!   IP address or through a relay using a TURN server), which are called "candidates". It then
+//!   generates a small text payload in a format called SDP, that describes the request for a
+//!   connection.
+//! - The offerer sends this SDP-encoded message to the answerer. The medium through which this
+//!   exchange is done is out of scope of the ICE protocol.
+//! - The answerer then finds its own candidates, and generates an answer, again in the SDP format.
+//!   This answer is sent back to the offerer.
+//! - Each agent then tries to connect to the remote's candidates.
+//!
+//! We pretend to send the offer to the remote agent (the target of the connection), then pretend
+//! that it has found a valid IP address for itself (i.e. a candidate), then pretend that the SDP
+//! answer containing this candidate has been sent back. This will cause the offerer to execute
+//! step 4: try to connect to the remote's candidate.
+//!
+//! ## 1 vs N data channels
+//!
+//! The SDP message generated by the offerer contains the list of so-called "media streams" that it
+//! wants to open. In our specific use-case, we configure the transport to always request one data
+//! stream.
+//!
+//! For this prototype, we have choosen to implement a single data channel and multiplex on top
+//! because doing N channels will force us to rewrite a significant part of smoldot, which isn't
+//! ready at all to have encryption and multiplexing handled "externally".
+//!
+//! ## TCP or UDP
+//!
+//! WebRTC by itself doesn't hardcode any specific protocol for these media streams. Instead, it is
+//! the SDP message of the offerer that specifies which protocol to use. In our use case, one data
+//! stream, we know that the offerer will always request either TCP+DTLS+SCTP, or UDP+DTLS+SCTP.
```

Comment:
Am I correct by assuming that this implementation is using UDP only? That sounds reasonable to me.

Reply:
That's correct. The code is built upon https://github.com/webrtc-rs/webrtc, which does not support UDP at the moment. It's possible that it will be supporting TCP in the future, in which case this code would need a few alterations.

Reply:
> which does not support UDP at the moment. 

Just to make sure we are on the same page, you meant to write "not support TCP", correct?

Reply:
> > which does not support UDP at the moment.
> 
> Just to make sure we are on the same page, you meant to write "not support TCP", correct?

Yes, sorry

Reply:
added a comment